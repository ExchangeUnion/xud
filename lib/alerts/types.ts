import { AlertType, ChannelSide } from '../constants/enums';

type BaseAlert = {
  type: AlertType;
  message: string;
  date: number;
};

export type BalanceAlertEvent = {
  /** The total balance of the channel when the alert is triggered. */
  totalBalance: number;
  /** The side of the balance either local or remote. */
  side: ChannelSide;
  /** The balance that triggered the alert. */
  sideBalance: number;
  /** The alert threshold in percentage, e.g. 10 means %10. */
  bound: number;
  /** The currency of the channel. */
  currency: string;
};
export type BalanceAlert = BaseAlert & BalanceAlertEvent;

export type Alert = BalanceAlert;
