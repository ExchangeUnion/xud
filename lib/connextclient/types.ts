/**
 * The configurable options for the connext client.
 */
export type ConnextClientConfig = {
  disable: boolean;
  host: string;
  port: number;
  webhookhost: string;
  webhookport: number;
};

/** General information about the state of this connext client. */
export type ConnextInfo = {
  status: string;
  error?: string;
  address?: string;
  chain?: string;
  version?: string;
};

/** Response for eth_blockNumber call  */
export type ConnextBlockNumberResponse = {
  result: string;
};

/** Response for eth_gasPrice call  */
export type EthproviderGasPriceResponse = {
  result: string;
};

type Balance = {
  amount: string[];
  to: string[];
};

/** Connext channel details */
export type ConnextChannelDetails = {
  assetIds: string[];
  balances: Balance[];
  channelAddress: string;
  merkleRoot: string;
  processedDepositsA: string[];
  processedDepositsB: string[];
  networkContext: {
    chainId: number;
    channelFactoryAddress: string;
    transferRegistryAddress: string;
    providerUrl: string;
  };
  nonce: number;
  alice: string;
  aliceIdentifier: string;
  bob: string;
  bobIdentifier: string;
  timeout: string;
  latestUpdate: {
    assetId: string;
    balance: Balance;
    channelAddress: string;
    details: {
      totalDepositsAlice: string;
      totalDepositsBob: string;
    };
    fromIdentifier: string;
    nonce: number;
    aliceSignature: string;
    bobSignature: string;
    toIdentifier: string;
    type: string;
  };
  defundNonce: string;
  inDispute: boolean;
};

/**
 * The payload for tokenPayment call.
 */
export type ConnextTransferRequest = {
  type: 'HashlockTransfer';
  channelAddress: string;
  amount: string;
  assetId: string;
  details: {
    lockHash: string;
    expiry: string;
  };
  recipient: string;
  meta: {
    routingId: string;
  };
  publicIdentifier: string;
};

/**
 * The response for errors.
 */
export type ConnextErrorResponse = { message: string };

export type ConnextConfig = {
  publicIdentifier: string;
  signerAddress: string;
  index: number;
};

/**
 * The response for /config call.
 */
export type ConnextConfigResponse = ConnextConfig[];

/**
 * The response for /channel call.
 */
export type ConnextChannelResponse = string[];

/**
 * The response for balance call.
 */
export type ConnextChannelBalanceResponse = {
  freeBalanceOffChain: string;
  nodeFreeBalanceOffChain: string;
  freeBalanceOnChain: string;
};

export type GetBlockByNumberResponse = {
  result: {
    difficulty: string;
    extraData: string;
    gasLimit: string;
    gasUsed: string;
    hash: string;
    logsBloom: string;
    miner: string;
    mixHash: string;
    nonce: string;
    number: string;
    parentHash: string;
    receiptsRoot: string;
    sha3Uncles: string;
    size: string;
    stateRoot: string;
    timestamp: string;
    totalDifficulty: string;
    transactions: string[];
    transactionsRoot: string;
    uncles: string[];
  };
};

/**
 * The response for hashLockTransfer call.
 */
export type ConnextTransferResponse = {
  channelAddress: string;
  transferId: string;
  routingId: string;
};

/**
 * The response for withdraw call.
 */
export type ConnextWithdrawResponse = {
  channelAddress: '0xa929dB0530daB525596f5d48Fb5C322fDa8A337B';
  transferId: '0xc2e4592d3fb6c02ee1d3b07bed83b5914f8ca084b0f91d6b80bf8107e58c9c38';
};

type ConnextRoutingPath = {
  recipient: string;
  recipientChainId: number;
  recipientAssetId: string;
};

export type ConnextTransfer = {
  channelFactoryAddress: string;
  assetId: string;
  chainId: number;
  channelAddress: string;
  balance: {
    amount: string[];
    to: string[];
  };
  initiator: string;
  responder: string;
  initialStateHash: string;
  transferDefinition: string;
  transferEncodings: string[];
  transferId: string;
  transferState: {
    lockHash: string;
    expiry: string;
  };
  transferTimeout: string;
  meta: {
    requireOnline: boolean;
    routingId: string;
    path: ConnextRoutingPath[];
    senderIdentifier?: string;
  };
  transferResolver?: {
    preImage?: string;
  };
};

export type TransfersByRoutingIdResponse = ConnextTransfer[];

export type ExpectedIncomingTransfer = {
  rHash: string;
  units: bigint;
  expiry: number;
  tokenAddress: string;
  routingId: string;
};

export type ConnextPreimageRequest = {
  channelAddress: string;
  channelBalance: {
    to: string[];
    amount: string[];
  };
  transfer: {
    channelFactoryAddress: string;
    assetId: string;
    chainId: string;
    channelAddress: string;
    balance: {
      amount: string[];
      to: string[];
    };
    initiator: string;
    responder: string;
    initialStateHash: string;
    transferDefinition: string;
    transferEncodings: string[];
    transferId: string;
    transferState: {
      lockHash: string;
      expiry: string;
    };
    transferTimeout: string;
    meta: {
      requireOnline: boolean;
      routingId: string;
      path: [
        {
          recipient: string;
          recipientChainId: number;
          recipientAssetId: string;
        },
      ];
    };
    transferResolver: {
      preImage: string;
    };
  };
  conditionType: string;
};

export type ConnextIncomingTransferRequest = {
  channelAddress: string;
  channelBalance: {
    to: string[];
    amount: string[];
  };
  transfer: {
    channelFactoryAddress: string;
    assetId: string;
    chainId: number;
    channelAddress: string;
    balance: {
      amount: string[];
      to: string[];
    };
    initiator: string;
    responder: string;
    initialStateHash: string;
    transferDefinition: string;
    transferEncodings: string[];
    transferId: string;
    transferState: {
      lockHash: string;
      expiry: string;
    };
    transferTimeout: string;
    meta: {
      routingId: string;
    };
  };
  conditionType: string;
};

export type ConnextDepositConfirmedRequest = {
  id: string;
  data?: {
    amount: {
      _hex: string;
    };
    assetId: string;
    hash: string;
  };
};

export type ProvidePreimageEvent = {
  rHash: string;
  preimage: string;
};

export type TransferReceivedEvent = {
  tokenAddress: string;
  rHash: string;
  expiry: number;
  units: bigint;
  routingId: string;
};
