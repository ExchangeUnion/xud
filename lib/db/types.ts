import { BelongsToGetAssociationMixin, Model } from 'sequelize';
import { ReputationEvent, SwapClientType } from '../constants/enums';
import { Currency, Order, Pair } from '../orderbook/types';
import { Address, NodeConnectionInfo } from '../p2p/types';
import { SwapDeal } from '../swaps/types';

/*
 * The following definitions are in sets of triplets, one for each Model (which represents a table in the database).
 *
 * "xCreationAttributes" is the type definition for the object which is required when a new record is to be created.
 *
 * "xAttributes" is the type definition of the record.
 *
 * "xInstance" is the type definition of a fetched record as a Sequelize row instance, which contains some util properties.
 */

/* Currency */
export type CurrencyCreationAttributes = Currency;

export type CurrencyAttributes = CurrencyCreationAttributes;

export interface CurrencyInstance extends Model<CurrencyAttributes, CurrencyCreationAttributes>, CurrencyAttributes {}

/* SwapDeal */
export type SwapDealCreationAttributes = Pick<
  SwapDeal,
  Exclude<keyof SwapDeal, 'takerMaxTimeLock' | 'price' | 'pairId' | 'isBuy' | 'takerUnits' | 'makerUnits'>
>;

export type SwapDealAttributes = SwapDealCreationAttributes & {
  /** The internal db node id of the counterparty peer for this swap deal. */
  nodeId: number;
};

export interface SwapDealInstance extends Model<SwapDealAttributes, SwapDealCreationAttributes>, SwapDealAttributes {
  getNode: BelongsToGetAssociationMixin<NodeInstance>;
  getOrder: BelongsToGetAssociationMixin<OrderInstance>;
  Node?: NodeInstance;
  Order?: OrderInstance;
}

export type OrderCreationAttributes = Pick<Order, Exclude<keyof Order, 'quantity' | 'hold' | 'price'>> & {
  /** The internal db node id of the peer that created this order. */
  nodeId?: number;
  localId?: string;
  /** The price for the order expressed in units of the quote currency. */
  price?: number;
};

export type OrderAttributes = OrderCreationAttributes & {};

export interface OrderInstance extends Model<OrderAttributes, OrderCreationAttributes>, OrderAttributes {}

export type TradeCreationAttributes = {
  /** The order id of the maker order involved in this trade. */
  makerOrderId: string;
  /** The order id of the taker order involved in this trade, if applicable. */
  takerOrderId?: string;
  /** The rHash of the swap that filled this trade, if applicable. */
  rHash?: string;
  /** The quantity transacted in this trade. */
  quantity: number;
};

export type TradeAttributes = TradeCreationAttributes & {};

export interface TradeInstance extends Model<TradeAttributes, TradeCreationAttributes>, TradeAttributes {
  getMakerOrder: BelongsToGetAssociationMixin<OrderInstance>;
  getTakerOrder: BelongsToGetAssociationMixin<OrderInstance>;
  getSwapDeal: BelongsToGetAssociationMixin<SwapDealInstance>;
  SwapDeal?: SwapDealInstance;
  makerOrder?: OrderInstance;
  takerOrder?: OrderInstance;
  createdAt: Date;
}

/* Node */
export type NodeCreationAttributes = NodeConnectionInfo;

export type NodeAttributes = NodeCreationAttributes & {
  id: number;
  banned: boolean;
  addressesText: string;
  lastAddressText: string;
  lastAddress: Address;
  consecutiveConnFailures: number;
};

export interface NodeInstance extends Model<NodeAttributes, NodeCreationAttributes>, NodeAttributes {
  reputationScore: number;
}

/* Pairs */
export type PairCreationAttributes = Pair;

export type PairAttributes = PairCreationAttributes & {
  id: string;
};

export interface PairInstance extends Model<PairAttributes, PairCreationAttributes>, PairAttributes {}

/* Reputation events */
export type ReputationEventCreationAttributes = {
  event: ReputationEvent;
  nodeId: number;
};

export type ReputationEventAttributes = ReputationEventCreationAttributes & {
  createdAt: number;
  id: number;
};

export interface ReputationEventInstance
  extends Model<ReputationEventAttributes, ReputationEventCreationAttributes>,
    ReputationEventAttributes {}

/* Passwords */
export type PasswordCreationAttributes = {
  encryptedPassword: string;
  currency?: string;
  swapClient: SwapClientType;
};

export type PasswordAttributes = PasswordCreationAttributes & {
  createdAt: number;
  id: number;
};

export interface PasswordInstance extends Model<PasswordAttributes, PasswordCreationAttributes>, PasswordAttributes {}
