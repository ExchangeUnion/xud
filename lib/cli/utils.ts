import colors from 'colors/safe';
import { accessSync, watch } from 'fs';
import os from 'os';
import path from 'path';

const SATOSHIS_PER_COIN = 10 ** 8;

export function getDefaultCertPath() {
  switch (os.platform()) {
    case 'win32': {
      const homeDir = process.env.LOCALAPPDATA!;
      return path.join(homeDir, 'Xud', 'tls.cert');
    }
    case 'darwin': {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
    default: {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
  }
}

export function shorten(key: string, length = 10) {
  if (key.length <= length * 2 + 3) {
    return key;
  }
  return `${key.slice(0, length)}...${key.slice(key.length - length)}`;
}

export function trim(key: string, length = 10) {
  if (key.length <= length + 3) {
    return '';
  }
  return `${key.slice(0, length)}...`;
}

export const generateHeaders = (headers: string[]) => {
  return headers.map((header) => {
    return colors.blue(header);
  });
};

/** Returns a number of coins as an integer number of satoshis. */
export const coinsToSats = (coinsQuantity: number) => {
  return Math.round(coinsQuantity * SATOSHIS_PER_COIN);
};

/** Returns a number of satoshis as a string representation of coins with up to 8 decimal places. */
export const satsToCoinsStr = (satsQuantity: number) => {
  return (satsQuantity / SATOSHIS_PER_COIN).toFixed(8).replace(/\.?0+$/, '');
};

/** Waits up to 5 seconds for the tls.cert file to be created in case this is the first time xud has been run. */
export const waitForCert = (certPath: string) => {
  return new Promise<void>((resolve, reject) => {
    try {
      accessSync(certPath);
      resolve();
    } catch (err) {
      if (err.code === 'ENOENT') {
        const certDir = path.dirname(certPath);
        const certFilename = path.basename(certPath);
        const fsWatcher = watch(certDir, (event, filename) => {
          if (event === 'change' && filename === certFilename) {
            clearTimeout(timeout);
            fsWatcher.close();
            resolve();
          }
        });
        const timeout = setTimeout(() => {
          fsWatcher.close();
          reject(`timed out waiting for cert to be created at ${certPath}`);
        }, 5000);
      } else {
        // we handle errors due to file not existing, otherwise reject
        reject(err);
      }
    }
  });
};

export function showSeed(seedMnemonicList: string[]) {
  const WORDS_PER_ROW = 4;
  const numberedMnemonic = seedMnemonicList.map((value, index) => {
    return `${index >= 9 ? '' : ' '}${index + 1}. ${value.padEnd(10)}`;
  });
  console.log('----------------------BEGIN XUD SEED---------------------');
  for (let n = 0; n < seedMnemonicList.length / WORDS_PER_ROW; n += 1) {
    console.log(numberedMnemonic.slice(n * WORDS_PER_ROW, (n + 1) * WORDS_PER_ROW).join(' '));
  }
  console.log('-----------------------END XUD SEED----------------------\n');

  console.log(`
Please write down your 24 word mnemonic. It will allow you to recover your xud \
node key and on-chain funds for the initialized wallets listed above should you \
forget your password or lose your device. Off-chain funds in channels can NOT \
be recovered with it and must be backed up and recovered separately. Keep it \
somewhere safe, it is your ONLY backup in case of data loss.
`);
}
