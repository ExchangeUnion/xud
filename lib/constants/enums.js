const enums = exports;

enums.swapProtocols = {
  LND: 'LND',
  RAIDEN: 'RAIDEN',
};

enums.orderingDirections = {
  DESC: 'DESC',
  ASC: 'ASC',
};
