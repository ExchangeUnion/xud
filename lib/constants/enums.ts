/** An enumeration of payment channel network clients that support token swaps. */
export enum SwapClients {
  Lnd,
  Raiden,
}

export enum OrderingDirection {
  Desc = 'DESC',
  Asc = 'ASC',
}

export enum OrderSide {
  Buy,
  Sell,
}

export enum Network {
  MainNet = 'mainnet',
  TestNet = 'testnet',
  SimNet = 'simnet',
  RegTest = 'regtest',
}

/**
 * Magic value indicating wire msg origin network, and used to seek to next msg when stream state is unknown
 */
export enum NetworkMagic {
  MainNet = 0xd9b4bef9,
  TestNet = 0x0709110b,
  SimNet = 0x12141c16,
  RegTest = 0xdab5bffa,
}

export enum SwapRole {
  Taker = 0,
  Maker = 1,
}

export enum SwapPhase {
  /** The swap has been created locally. */
  SwapCreated = 0,
  /** The swap has been sent to a peer to request approval. */
  SwapRequested = 1,
  /** The terms of the swap have been agreed to, and we will attempt to execute it. */
  SwapAgreed = 2,
  /**
   * We have commanded lnd to send payment according to the agreed terms. The payment (and swap)
   * could still fail due to no route with sufficient capacity, lack of cooperation from the
   * receiver or any intermediary node along the route, or an unexpected error from lnd.
   */
  SendingAmount = 3,
  /** We have received the agreed amount of the swap, and the preimage is now known to both sides. */
  AmountReceived = 4,
  /** The swap has been formally completed and both sides have confirmed they've received payment. */
  SwapCompleted = 5,
}

export enum SwapState {
  Active = 0,
  Error = 1,
  Completed = 2,
}

export enum ReputationEvent {
  ManualBan = 0,
  ManualUnban = 1,
  PacketTimeout = 2,
  WireProtocolErr = 3,
  InvalidAuth = 4,
  SwapSuccess = 5,
  /** When a swap is accepted and is attempted to be executed but fails. */
  SwapFailure = 6,
  /** When a swap is accepted and then fails due to exceeding time limits. */
  SwapTimeout = 7,
  /** When a swap fails due to unexpected or possibly malicious behavior. */
  SwapMisbehavior = 8,
}

export enum SwapFailureReason {
  /** Could not find the order specified by a swap request. */
  OrderNotFound = 0,
  /** The order specified by a swap request is on hold for a different ongoing swap. */
  OrderOnHold = 1,
  /** The swap request contained invalid data. */
  InvalidSwapRequest = 2,
  /** We are not connected to both swap clients, or we are missing pub key identifiers for the peer's nodes. */
  SwapClientNotSetup = 3,
  /** Could not find a route to complete the swap. */
  NoRouteFound = 4,
  /** A call to lnd failed for an unexpected reason. */
  UnexpectedLndError = 5,
  /** Received a swap packet from the peer with invalid data. */
  InvalidSwapPacketReceived = 6,
  /** The call to lnd to send payment failed. */
  SendPaymentFailure = 7,
  /** The swap resolver request from lnd was invalid. */
  InvalidResolveRequest = 8,
  /** The swap request attempts to reuse a payment hash. */
  PaymentHashReuse = 9,
  /** The swap timed out while we were waiting for it to complete execution. */
  SwapTimedOut = 10,
  /** The deal timed out while we were waiting for the peer to respond to our swap request. */
  DealTimedOut = 11,
  /** The swap failed due to an unrecognized error. */
  UnknownError = 12,
}

export enum DisconnectionReason {
  ResponseStalling = 1,
  IncompatibleProtocolVersion = 2,
  UnexpectedIdentity = 3,
  ForbiddenIdentityUpdate = 4,
  ConnectedToSelf = 5,
  NotAcceptingConnections = 6,
  Banned = 7,
  AlreadyConnected = 8,
  Shutdown = 9,
  MalformedVersion = 10,
  AuthFailureInvalidTarget = 11,
  AuthFailureInvalidSignature = 12,
  WireProtocolErr = 13,
}
