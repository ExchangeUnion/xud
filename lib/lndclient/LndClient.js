/* eslint-disable camelcase */

const grpc = require('grpc');
const fs = require('fs');
const assert = require('assert');

/** A class representing a client to interact with a running lnd instance. */
class LndClient {
  /**
   * Create an lnd client.
   * @param {Object} options - The lnd configuration
   */
  constructor(options) {
    const { datadir, rpcProtoPath } = options;
    assert(typeof options.datadir === 'string', 'datadir must be a string');
    assert(typeof options.rpcProtoPath === 'string', 'rpcProtoPath must be a string');

    const lndCert = fs.readFileSync(`${datadir}tls.cert`);
    const credentials = grpc.credentials.createSsl(lndCert);
    const lnrpcDescriptor = grpc.load(rpcProtoPath);
    this.lightning = new lnrpcDescriptor.lnrpc.Lightning('127.0.0.1:10009', credentials);

    const adminMacaroon = fs.readFileSync(`${datadir}admin.macaroon`);
    this.meta = new grpc.Metadata();
    this.meta.add('macaroon', adminMacaroon.toString('hex'));
  }

  /**
   * Return general information concerning the lightning node including itâ€™s identity pubkey,
   * alias, the chains it is connected to, and information concerning the number of
   * open+pending channels.
   */
  getInfo() {
    return new Promise((resolve, reject) => {
      this.lightning.getInfo({}, this.meta, (err, response) => {
        if (err) {
          reject(err.details);
        } else {
          resolve(response);
        }
      });
    });
  }

  /**
   * Attempt to add a new invoice to the lnd invoice database.
   * @param {number} value - The value of this invoice in satoshis
   */
  addInvoice(value) {
    return new Promise((resolve, reject) => {
      this.lightning.addInvoice({
        value,
        memo: 'XU',
      }, this.meta, (err, response) => {
        if (err) {
          reject(err.details);
        } else {
          resolve(response);
        }
      });
    });
  }

  /**
   * Pay an invoice through the Lightning Network.
   * @param {string} payment_request - An invoice for a payment within the Lightning Network.
   */
  payInvoice(payment_request) {
    return new Promise((resolve, reject) => {
      this.lightning.sendPaymentSync({ payment_request }, this.meta, (err, response) => {
        if (err) {
          reject(err.details);
        } else {
          resolve(response);
        }
      });
    });
  }
}

module.exports = LndClient;
