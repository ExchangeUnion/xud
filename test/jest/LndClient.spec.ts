import LndClient from '../../lib/lndclient/LndClient';
import { LndClientConfig } from '../../lib/lndclient/types';
import Logger from '../../lib/Logger';
import { getValidDeal } from '../utils';
import { SwapRole } from '../../lib/constants/enums';
import { ClientStatus } from '../../lib/swaps/SwapClient';

const openChannelSyncResponse = {
  hasFundingTxidStr: () => {
    return true;
  },
  getFundingTxidStr: () => 'some_tx_id',
};

const preimage = 'IDAKXrx4dayn0H/gCxN12jPK2/LchwPZop4zICw43jg=';

jest.mock('../../lib/Logger');
const mockedLogger = <jest.Mock<Logger>>(<any>Logger);
describe('LndClient', () => {
  let lnd: LndClient;
  let config: LndClientConfig;
  let currency: string;
  let logger: Logger;

  beforeEach(() => {
    config = {
      disable: false,
      certpath: '/cert/path',
      macaroonpath: '/macaroon/path',
      host: '127.0.0.1',
      port: 4321,
      nomacaroons: true,
      cltvdelta: 40,
    };
    currency = 'BTC';
    logger = new mockedLogger();
    logger.error = jest.fn();
    logger.info = jest.fn();
    logger.trace = jest.fn();
    logger.debug = jest.fn();

    lnd = new LndClient({ config, currency, logger });
    lnd['status'] = ClientStatus.ConnectionVerified;
  });

  afterEach(async () => {
    jest.clearAllMocks();
    lnd.close();
  });

  describe('openChannel', () => {
    const peerPubKey = '02f8895eb03c37b2665415be4d83b20228acc0abc55ebf6728565141c66cfc164a';
    const units = 16000000n;
    const externalIp1 = '123.456.789.321:9735';
    const externalIp2 = '192.168.63.155:9777';
    const lndListeningUris = [`${peerPubKey}@${externalIp1}`, `${peerPubKey}@${externalIp2}`];

    test('it tries all 2 lnd uris when connectPeer to first one fails', async () => {
      expect.assertions(3);
      lnd['openChannelSync'] = jest.fn().mockReturnValue(Promise.resolve(openChannelSyncResponse));
      const connectPeerFail = () => {
        throw new Error('connectPeer failed');
      };
      lnd['connectPeer'] = jest
        .fn()
        .mockImplementationOnce(connectPeerFail)
        .mockImplementationOnce(() => {
          return Promise.resolve();
        });
      await lnd.openChannel({
        units,
        remoteIdentifier: peerPubKey,
        uris: lndListeningUris,
      });
      expect(lnd['connectPeer']).toHaveBeenCalledTimes(2);
      expect(lnd['connectPeer']).toHaveBeenCalledWith(peerPubKey, externalIp1);
      expect(lnd['connectPeer']).toHaveBeenCalledWith(peerPubKey, externalIp2);
    });

    test('it does succeed when connecting to already connected peer', async () => {
      expect.assertions(4);
      lnd['openChannelSync'] = jest.fn().mockReturnValue(Promise.resolve(openChannelSyncResponse));
      const alreadyConnected = () => {
        throw new Error('already connected');
      };
      lnd['connectPeer'] = jest.fn().mockImplementation(alreadyConnected);
      await lnd.openChannel({
        units,
        remoteIdentifier: peerPubKey,
        uris: lndListeningUris,
      });
      expect(lnd['connectPeer']).toHaveBeenCalledTimes(1);
      expect(lnd['connectPeer']).toHaveBeenCalledWith(peerPubKey, externalIp1);
      expect(lnd['openChannelSync']).toHaveBeenCalledTimes(1);
      expect(lnd['openChannelSync']).toHaveBeenCalledWith(peerPubKey, Number(units), 0, 0);
    });

    test('it pushes satoshis to the peer when specified', async () => {
      expect.assertions(4);
      const pushUnits = 481824n;
      lnd['openChannelSync'] = jest.fn().mockReturnValue(Promise.resolve(openChannelSyncResponse));
      lnd['connectPeer'] = jest.fn().mockImplementationOnce(() => {
        return Promise.resolve();
      });
      await lnd.openChannel({
        units,
        pushUnits,
        remoteIdentifier: peerPubKey,
        uris: lndListeningUris,
      });
      expect(lnd['connectPeer']).toHaveBeenCalledTimes(1);
      expect(lnd['connectPeer']).toHaveBeenCalledWith(peerPubKey, externalIp1);
      expect(lnd['openChannelSync']).toHaveBeenCalledTimes(1);
      expect(lnd['openChannelSync']).toHaveBeenCalledWith(peerPubKey, Number(units), Number(pushUnits), 0);
    });

    test('it should set fee when specified', async () => {
      expect.assertions(4);
      const fee = 24;
      lnd['openChannelSync'] = jest.fn().mockReturnValue(Promise.resolve(openChannelSyncResponse));
      lnd['connectPeer'] = jest.fn().mockImplementationOnce(() => {
        return Promise.resolve();
      });
      await lnd.openChannel({
        units,
        fee,
        remoteIdentifier: peerPubKey,
        uris: lndListeningUris,
      });
      expect(lnd['connectPeer']).toHaveBeenCalledTimes(1);
      expect(lnd['connectPeer']).toHaveBeenCalledWith(peerPubKey, externalIp1);
      expect(lnd['openChannelSync']).toHaveBeenCalledTimes(1);
      expect(lnd['openChannelSync']).toHaveBeenCalledWith(peerPubKey, Number(units), 0, fee);
    });

    test('it stops trying to connect to lnd uris when first once succeeds', async () => {
      expect.assertions(3);
      lnd['openChannelSync'] = jest.fn().mockReturnValue(Promise.resolve(openChannelSyncResponse));
      lnd['connectPeer'] = jest.fn().mockImplementationOnce(() => {
        return Promise.resolve();
      });
      await lnd.openChannel({
        units,
        remoteIdentifier: peerPubKey,
        uris: lndListeningUris,
      });
      expect(lnd['connectPeer']).toHaveBeenCalledTimes(1);
      expect(lnd['connectPeer']).toHaveBeenCalledWith(peerPubKey, externalIp1);
      expect(lnd['connectPeer']).not.toHaveBeenCalledWith(peerPubKey, externalIp2);
    });

    test('it throws when openchannel fails', async () => {
      expect.assertions(2);
      lnd['connectPeer'] = jest.fn().mockReturnValue(Promise.resolve());
      lnd['openChannelSync'] = jest.fn().mockImplementation(() => {
        throw new Error('openChannelSync error');
      });
      try {
        await lnd.openChannel({
          units,
          remoteIdentifier: peerPubKey,
          uris: lndListeningUris,
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
      expect(lnd['openChannelSync']).toHaveBeenCalledTimes(1);
    });
  });

  describe('sendPayment', () => {
    test('it resolves upon maker success', async () => {
      lnd['sendPaymentV2'] = jest.fn().mockReturnValue(Promise.resolve(preimage));
      const deal = getValidDeal();
      const buildSendRequestSpy = jest.spyOn(lnd as any, 'buildSendRequest');
      await expect(lnd.sendPayment(deal)).resolves.toEqual(preimage);
      expect(buildSendRequestSpy).toHaveBeenCalledWith({
        amount: deal.takerAmount,
        destination: deal.takerPubKey,
        rHash: deal.rHash,
        cltvLimit: deal.takerMaxTimeLock + 3,
        finalCltvDelta: deal.takerCltvDelta,
      });
    });

    test('it resolves upon taker success', async () => {
      lnd['sendPaymentV2'] = jest.fn().mockReturnValue(Promise.resolve(preimage));
      const deal = {
        ...getValidDeal(),
        role: SwapRole.Taker,
      };
      const buildSendRequestSpy = jest.spyOn(lnd as any, 'buildSendRequest');
      await expect(lnd.sendPayment(deal)).resolves.toEqual(preimage);
      expect(buildSendRequestSpy).toHaveBeenCalledWith({
        amount: deal.makerAmount,
        destination: deal.destination,
        finalCltvDelta: deal.makerCltvDelta,
        rHash: deal.rHash,
      });
    });

    test('it rejects upon sendPaymentSync error', async () => {
      lnd['sendPaymentV2'] = jest.fn().mockRejectedValue('error');
      await expect(lnd.sendPayment(getValidDeal())).rejects.toEqual('error');
    });

    test('it resolves upon sendSmallestAmount success', async () => {
      lnd['sendPaymentV2'] = jest.fn().mockReturnValue(Promise.resolve(preimage));
      const buildSendRequestSpy = jest.spyOn(lnd as any, 'buildSendRequest');
      const rHash = '04b6ac45b770ec4abbb9713aebfa57b963a1f6c7a795d9b5757687e0688add80';
      const destination = '034c5266591bff232d1647f45bcf6bbc548d3d6f70b2992d28aba0afae067880ac';
      await expect(lnd.sendSmallestAmount(rHash, destination)).resolves.toEqual(preimage);
      expect(buildSendRequestSpy).toHaveBeenCalledWith({
        destination,
        rHash,
        finalCltvDelta: lnd.finalLock,
        amount: 1,
      });
    });
  });

  describe('swapCapacities', () => {
    test('fetch and persist swap capacities', async () => {
      expect.assertions(5);

      lnd['pendingChannels'] = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          toObject: () => {
            return { pendingOpenChannelsList: [] };
          },
        });
      });

      lnd['listChannels'] = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          toObject: () => {
            return {
              channelsList: [
                {
                  active: true,
                  localBalance: 100,
                  localChanReserveSat: 2,
                  remoteBalance: 200,
                  remoteChanReserveSat: 5,
                },
                {
                  active: true,
                  localBalance: 80,
                  localChanReserveSat: 2,
                  remoteBalance: 220,
                  remoteChanReserveSat: 5,
                },
                {
                  active: true,
                  localBalance: 110,
                  localChanReserveSat: 20,
                  remoteBalance: 300,
                  remoteChanReserveSat: 5,
                },
                {
                  active: false,
                  localBalance: 50,
                  localChanReserveSat: 2,
                  remoteBalance: 50,
                  remoteChanReserveSat: 2,
                },
              ],
            };
          },
        });
      });

      const swapCapacities = await lnd.swapCapacities();
      expect(swapCapacities.maxOutboundChannelCapacity).toEqual(98);
      expect(swapCapacities.maxInboundChannelCapacity).toEqual(295);

      expect(lnd['listChannels']).toHaveBeenCalledTimes(1);
      expect(lnd['maxChannelOutboundAmount']).toEqual(98);
      expect(lnd['maxChannelInboundAmount']).toEqual(295);
    });
  });
});
