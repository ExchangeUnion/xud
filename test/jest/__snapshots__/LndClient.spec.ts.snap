// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LndClient openChannel it throws when openchannel fails 1`] = `[Error: openChannelSync error]`;
